rust/src/services/messages/message_parser.rs

```diff
@@ -0,0 +1,228 @@
+        Sum2Owned,
+        SumOwned,
+        Tag,
+        ToBytes,
+        UpdateOwned,
+    },
+    state_machine::{
+        coordinator::PhaseEvent,
+        events::{EventListener, EventSubscriber},
+    },
+    utils::trace::{Traceable, Traced},
+    Signature,
+};
+
+pub struct MessageParserService {
+    keys_events: EventListener<KeyPair>,
+    phase_events: EventListener<PhaseEvent>,
+
+    thread_pool: Arc<ThreadPool>,
+}
+
+impl MessageParserService {
+    pub fn new(subscriber: &EventSubscriber, thread_pool: Arc<ThreadPool>) -> Self {
+        Self {
+            keys_events: subscriber.keys_listener(),
+            phase_events: subscriber.phase_listener(),
+            thread_pool,
+        }
+    }
+}
+
+#[derive(From, Debug)]
+pub struct MessageParserRequest(Vec<u8>);
+
+pub type MessageParserResponse = Result<MessageOwned, MessageParserError>;
+
+#[derive(Debug, Error)]
+pub enum MessageParserError {
+    #[error("Failed to decrypt the message with the coordinator secret key")]
+    Decrypt,
+
+    #[error("Parsing failed: {0:?}")]
+    Parsing(DecodeError),
+
+    #[error("Invalid message signature")]
+    InvalidMessageSignature,
+
+    #[error("The message was rejected because the coordinator did not expect it")]
+    UnexpectedMessage,
+
+    // TODO: we should have a retry layer that automatically retries
+    // requests that fail with this error.
+    #[error("The request could not be processed due to a temporary internal error")]
+    TemporaryInternalError,
+
+    #[error("Internal error: {0}")]
+    InternalError(String),
+}
+
+impl Service<Traced<MessageParserRequest>> for MessageParserService {
+    type Response = MessageParserResponse;
+    type Error = std::convert::Infallible;
+    type Future = Either<
+        future::Ready<Result<Self::Response, Self::Error>>,
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + 'static + Send + Sync>>,
+    >;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: Traced<MessageParserRequest>) -> Self::Future {
+        debug!("retrieving the current keys and current phase");
+        let keys_ev = self.keys_events.get_latest();
+        let phase_ev = self.phase_events.get_latest();
+
+        // This can happen if the coordinator is switching starting a
+        // new phase. The error should be temporary and we should be
+        // able to retry the request.
+        if keys_ev.round != phase_ev.round {
+            return Either::Left(future::ready(Ok(Err(
+                MessageParserError::TemporaryInternalError,
+            ))));
+        }
+
+        let handler = Handler {
+            keys: keys_ev.event,
+            phase: phase_ev.event,
+        };
+
+        let (tx, rx) = oneshot::channel::<Self::Response>();
+
+        trace!("spawning pre-processor handler on thread-pool");
+        self.thread_pool.spawn(move || {
```

Comment:
Using `rayon` here is really cool. I also had in mind that we could use `tokio::spawn_blocking` but then I saw that they say in the tokio documentation:

> If your code is CPU-bound and you wish to limit the number of threads used to run it, you should run it on another thread pool such as rayon. You can use an oneshot channel to send the result back to Tokio when the rayon task finishes.

So `rayon` is the way to go. However, I don't really know how `tokio` and `rayon` split up the cpu cores among each other. So the issue I see here is that `tokio` and `rayon` run on the same core and interfere each other.

Reply:
Good question. I don't know =/

Reply:
I can do some research :)
Already found something https://docs.rs/tokio/0.2.21/tokio/#cpu-bound-tasks-and-blocking-code